core/optaplanner-constraint-streams-bavet/src/main/java/org/optaplanner/constraint/streams/bavet/common/AbstractGroupNode.java

```diff
@@ -123,124 +153,71 @@ public void update(InTuple_ tuple) {
-                        oldGroup.outTuple.setState(BavetTupleState.ABORTING);
-                    }
-                    break;
-                case UPDATING:
-                    if (killGroup) {
-                        oldGroup.outTuple.setState(BavetTupleState.DYING);
-                    }
-                    break;
-                case OK:
-                    oldGroup.outTuple.setState(killGroup ? BavetTupleState.DYING : BavetTupleState.UPDATING);
-                    dirtyGroupQueue.add(oldGroup);
-                    break;
-                case DYING:
-                case ABORTING:
-                case DEAD:
-                default:
-                    throw new IllegalStateException("Impossible state: The group (" + oldGroup + ") in node (" +
-                            this + ") is in an unexpected state (" + oldGroup.outTuple.getState() + ").");
-            }
-
-            Group<OutTuple_, GroupKey_, ResultContainer_> newGroup = groupMap.get(newGroupKey);
-            if (newGroup == null) {
-                ResultContainer_ resultContainer = hasCollector ? supplier.get() : null;
-                OutTuple_ outTuple = createOutTuple(newGroupKey);
-                outTuple.setState(BavetTupleState.CREATING);
-                newGroup = new Group<>(newGroupKey, resultContainer, outTuple);
-                groupMap.put(newGroupKey, newGroup);
-                // Don't add it if (state == CREATING), but (newGroup != null), which is a 2th insert of the same newGroupKey.
-                dirtyGroupQueue.add(newGroup);
-            }
-            newGroup.parentCount++;
-            Runnable undoAccumulator = hasCollector ? accumulate(newGroup.resultContainer, tuple) : null;
-            GroupPart<Group<OutTuple_, GroupKey_, ResultContainer_>> newGroupPart = new GroupPart<>(newGroup, undoAccumulator);
-            tupleStore[groupStoreIndex] = newGroupPart;
-
-            switch (newGroup.outTuple.getState()) {
-                case CREATING:
-                    break;
-                case UPDATING:
-                    break;
-                case OK:
-                    newGroup.outTuple.setState(BavetTupleState.UPDATING);
-                    dirtyGroupQueue.add(newGroup);
-                    break;
-                case DYING:
-                    newGroup.outTuple.setState(BavetTupleState.UPDATING);
-                    break;
-                case ABORTING:
-                    newGroup.outTuple.setState(BavetTupleState.CREATING);
-                    break;
-                case DEAD:
-                default:
-                    throw new IllegalStateException("Impossible state: The group (" + newGroup + ") in node (" +
-                            this + ") is in an unexpected state (" + newGroup.outTuple.getState() + ").");
-            }
+            killTuple(oldGroup);
+            createTuple(tuple, tupleStore, newGroupKey);
         }
     }
 
-    @Override
-    public void retract(InTuple_ tuple) {
```

Comment:
Nitpick: any reason the retract method is moved under the killTuple() method it delegates too?
Philosophy: reading from the top-down, I expect to insert/update/retract to be nearer to the top and killtuple to be more down (a detail of calls to the method above)

Reply:
`killTuple()` is also used in `update()`, which comes first. Therefore this method comes second. The order is `update`, `killTuple`, `retract`.